slicedimage/backends/_base.py

```diff
@@ -16,5 +19,37 @@ def write_file_from_handle(self, name, source_handle, block_size=(128 * 1024)):
             dest_handle.write(data)
 
 
-class FileNotFoundError(Exception):
-    pass
+class ChecksummingContextManager(object):
+    def __init__(self, wrapped_file_contextmanager, expected_sha256_checksum):
+        self._wrapped_file_contextmanager = wrapped_file_contextmanager
+        self._io = None
+        self._checksummer = hashlib.sha256()
+        self._expected_sha256_checksum = expected_sha256_checksum
+
+    def __enter__(self):
+        self._io = ChecksummingFile(
+            self._wrapped_file_contextmanager.__enter__(), self._checksummer)
+        return self._io
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # finish reading the data.
+        self._io.read()
+        calculated_checksum = self._checksummer.hexdigest()
+        if (self._expected_sha256_checksum is not None and
+                calculated_checksum != self._expected_sha256_checksum):
+            raise ValueError(
+                "calculated checksum ({}) does not match expected checksum ({})".format(
+                    calculated_checksum, self._expected_sha256_checksum))
+
+        return self._wrapped_file_contextmanager.__exit__(exc_type, exc_val, exc_tb)
+
+
+class ChecksummingFile(io.IOBase):
```

Comment:
Since this is called by the ContextManager, I'd move it above. If it's only used in the ContextManager, could this be made private? 

Similarly, could we doc this to describe its uses?

Reply:
Done.